### DISCUSSION TITLE
CharacterVirtual experiences jittery behavior when jumping in obtuse corners

### DISCUSSION THREAD

**Discussion by a-day-old-bagel** [jrouwe/JoltPhysics#1044] [Q&A] [answered]
I don't know if this is just an inherent drawback of how CharacterVirtual is designed, or if it can be alleviated by tweaking the CharacterVirtual settings (I've tried tweaking them to no avail), or if this is a bug that can be tracked down.

I've been experiencing an issue in my own implementation using the virtual character controller, and I noticed the same behavior is evident in the sample. When jumping against an inside corner (where two walls meet), continually pressing the forward button while jumping, some corners (those that meet at an obtuse angle) cause the character controller to jitter around.

Right angles or acute angles seem to behave just fine. It seems like a minor issue, but it's causing some problems with my gameplay. I'll include a video of some corners that have the issue and others that don't:

https://github.com/jrouwe/JoltPhysics/assets/12072151/fe56da04-3828-4d61-96d8-6a193984732e

**Comment by jrouwe** [comment, marked as answer]
I'm aware of this behavior. It is a result of the sliding code:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/bbfdf145-bdf9-4dc4-86a5-3f0e23944d98)

The picture above shows the situation from above, in 1 frame it is possible that the character follows the indicated path to slide along both surfaces. The next frame, the user would still press the forward key and it would slide back again over the same 2 surfaces, causing an oscillation.

Currently the only condition that stops sliding is if the velocity reverses when it hits a surface:

https://github.com/jrouwe/JoltPhysics/blob/5c9d3b011ab76cdaa05c9e1b1e699f716bd1ddc4/Jolt/Physics/Character/CharacterVirtual.cpp#L749

but that's not happening here since it changes velocity 2x. I remember playing around with this a lot and getting to this compromise, but I don't remember anymore what the downsides of other approaches were.

**Reply by a-day-old-bagel** [reply]
Thanks, that makes sense. Maybe I'll have to try some higher-level workarounds in my use case.

**Comment by miketuritzin** [comment]
I wanted to add that I'm seeing jittering that could be due to the same (or a similar) issue with `CharacterVirtual`, which I recently integrated in my engine. Generally it is working pretty well but for some geometry I'm seeing a lot of jittering.

Note that I'm generating collision geometry using marching cubes from an SDF, so the collision mesh isn't super clean.

In this video, I'm walking directly forward several times, and you can see the jittering (collision mesh rendering is enabled):

https://github.com/user-attachments/assets/31c8f636-0ff9-4d32-981c-79c0eb70d23c

My setup is identical to `CharacterVirtualTest` (and am using a capsule collider), and I've tried adjusting a lot of parameters, including setting `mEnhancedInternalEdgeRemoval ` to `true` and dramatically decreasing `mPredictiveContactDistance` (as suggested [here](https://github.com/jrouwe/JoltPhysics/discussions/1695)), but nothing has really had an impact so far.

I'll keep messing with this, but any tips appreciated!

**Reply by jrouwe** [reply]
I can't really tell from this video. This would need some sort of repro in the Samples app (you can save a binary snapshot of your world, see the LoadBinarySnapshot test) and it will take a couple of weeks before I can look at it.

**Reply by miketuritzin** [reply]
Gotcha, I will try to put something together. Thanks for the response.

**Reply by miketuritzin** [reply]
Okay, I've created a repro in `CharacterVirtualTest` that loads a snapshot saved from my engine. The `CharacterVirtual` is spawned higher up so it is above the terrain that gets added in.

See this commit in my fork that contains the snapshot and code changes to load it:
https://github.com/miketuritzin/JoltPhysicsCharacterVirtualJitterRepro/commit/2aab403262d740f82f73812e7196a57cf6e9b476

This video demonstrates one case of the jittering:

https://github.com/user-attachments/assets/4de44ef0-d5d6-4756-ba59-23c3d7c4bc3b

Note that the terrain is formed from 4 chunks that share edges, but this seems unrelated to the issue, as it happens when fully inside a single chunk.

Appreciate it if you have time to take a look at this.

**Reply by jrouwe** [reply]
Small update: I can reproduce the issue and I had a fix for the issue. Unfortunately it completely broke character interaction with moving platforms, so now it's back to the drawing board...

**Reply by miketuritzin** [reply]
Good to hear you're taking a look. In case you end up not getting to a fix, I am curious to see the previous one that broke moving platforms, as I don't use them :)

**Reply by jrouwe** [reply]
```
diff --git a/Jolt/Physics/Character/CharacterVirtual.cpp b/Jolt/Physics/Character/CharacterVirtual.cpp
index 40679f88..7c12f9b1 100644
--- a/Jolt/Physics/Character/CharacterVirtual.cpp
+++ b/Jolt/Physics/Character/CharacterVirtual.cpp
@@ -775,9 +775,6 @@ void CharacterVirtual::SolveConstraints(Vec3Arg inVelocity, float inDeltaTime, f
 	// This is the velocity we use for the displacement, if we hit something it will be shortened
 	Vec3 velocity = inVelocity;
 
-	// Keep track of the last velocity that was applied to the character so that we can detect when the velocity reverses
-	Vec3 last_velocity = inVelocity;
-
 	// Start with no displacement
 	outDisplacement = Vec3::sZero();
 	outTimeSimulated = 0.0f;
@@ -1002,9 +999,7 @@ void CharacterVirtual::SolveConstraints(Vec3Arg inVelocity, float inDeltaTime, f
 			return;
 
 		// If the constraint has velocity we accept the new velocity, otherwise check that we didn't reverse velocity
-		if (!constraint->mLinearVelocity.IsNearZero(1.0e-8f))
-			last_velocity = constraint->mLinearVelocity;
-		else if (velocity.Dot(last_velocity) < 0.0f)
+		if (constraint->mLinearVelocity.IsNearZero(1.0e-8f) && velocity.Dot(inVelocity) < 0.0f)
 			return;
 	}
 }
```
